link/syscalls.go

```diff
@@ -127,18 +127,32 @@ var haveProgQuery = internal.NewFeatureTest("BPF_PROG_QUERY", "4.15", func() err
 })
 
 var haveTCX = internal.NewFeatureTest("tcx", "6.6", func() error {
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Type:    ebpf.SchedCLS,
+		License: "MIT",
+		Instructions: asm.Instructions{
+			asm.Mov.Imm(asm.R0, 0),
+			asm.Return(),
+		},
+	})
+
+	if err != nil {
+		return internal.ErrNotSupported
+	}
+
+	defer prog.Close()
 	attr := sys.LinkCreateTcxAttr{
 		// We rely on this being checked during the syscall.
-		// With an otherwise correct payload we expect EBADF here
+		// With an otherwise correct payload we expect ENODEV here
 		// as an indication that the feature is present.
-		ProgFd:        ^uint32(0),
 		TargetIfindex: ^uint32(0),
+		ProgFd:        uint32(prog.FD()),
 		AttachType:    sys.AttachType(ebpf.AttachTCXIngress),
```

Comment:
The reason is that AttachType is checked after ifindex / progfd is checked?

Reply:
First the program type is checked, then the AttachType. See https://elixir.bootlin.com/linux/v6.7/source/kernel/bpf/syscall.c#L4991. `tcx_link_attach` will then return `ENODEV` because of the invalid ifindex. On kernels without tcx support `EINVAL` will be returned because `BPF_LINK_CREATE` doesn't know how to create links for programs with type `SchedCLS`.